### PR TITLE
Retry Zcash sprout and sapling parameters download

### LINKED ISSUE #3239: Add retries for the sapling and sprout parameter downloads
## Motivation

A build of PR #3191 failed with:
> Dec 15 21:55:10.404  INFO {zebrad="de6b673" net="Main"}: zebra_consensus::primitives::groth16::params: downloading Zcash Sapling parameters
> Dec 15 21:56:15.162  INFO {zebrad="de6b673" net="Main"}: zebra_consensus::primitives::groth16::params: downloading Zcash Sprout parameters
> The application panicked (crashed).
> Message:  error downloading Sprout parameter files: IoError(Custom { kind: InvalidData, error: "sprout-groth16.params failed validation: expected: e9b238411bd6c0ec4791e9d04245ec350c9c5744f5610dfcce4365d5ca49dfefd5054e371842b3f88fa1b9d7e8e075249b3ebabd167fa8b0f3161292d36c180a hashing 725523612 bytes, actual: 980ecfb84bef129ff5547a5462e29c2cbff91ecf717d17aec15c6d506a12297e28a8b1c7cac9e3faddce712d02575980e8d29ee232bba3e32f438c14c256e69f hashing 609221143 bytes from \"https://download.z.cash/downloads/sprout-groth16.params.part.1 + https://download.z.cash/downloads/sprout-groth16.params.part.2\"" }). Hint: try deleting "/root/.zcash-params", then running 'zebrad download' to re-download the parameters
Location: zebra-consensus/src/primitives/groth16/params.rs:88

### Designs

- Add a retry to Zebra's parameter downloads
- Only retry the downloads that actually failed

https://github.com/ZcashFoundation/zebra/blob/ea01e305ed0205df7ac47ff757a2036fef576fd5/zebra-consensus/src/primitives/groth16/params.rs#L71

https://github.com/ZcashFoundation/zebra/blob/ea01e305ed0205df7ac47ff757a2036fef576fd5/zebra-consensus/src/primitives/groth16/params.rs#L86

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to retry a few times the Sprout and Sapling parameter download if they fail for whatever reason.

Fixes https://github.com/ZcashFoundation/zebra/issues/3239

## Solution

Retry the downloads if they fail. Number of retries is defined by a constant(currently = 3).

The code can be maybe improved a bit to make it shorter but i think it should work as it is too.

## Review

Anyone can review, the code is pretty simple.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-consensus/src/primitives/groth16/params.rs [hunk 1]
```diff
@@ -67,31 +73,56 @@ impl Groth16Parameters {
...
-            assert_eq!(sapling_output_path, new_sapling_paths.output);
+
+            let mut sapling_retries = 0;
+            while sapling_retries < PARAMETER_DOWNLOAD_MAX_RETRIES {
+                let new_sapling_paths =
+                    zcash_proofs::download_sapling_parameters(Some(PARAMETER_DOWNLOAD_TIMEOUT));
+
+                if new_sapling_paths.is_err() {
+                    sapling_retries += 1;
+                    tracing::info!("error downloading Zcash Sapling parameters, retrying");
```
Comment (oxarbitrage) [inline zebra-consensus/src/primitives/groth16/params.rs]: I tried to put the underlying error in the message but i had compiler issues as:

```
 doesn't satisfy `SaplingParameterPaths: std::fmt::Debug`
 ```

where deriving `Debug` is not an option as it is outside our code. I think there should be a way to do this in our side but i was not able to find how.

Comment (teor2345) [reply zebra-consensus/src/primitives/groth16/params.rs]: I actually wrote that code, and my `zcash_proofs` PR is still waiting for review, so it was easy to fix.

You can get those changes in Zebra once the dependency update in #3308 merges.

**Review by teor2345 — COMMENTED** [review]
Logging the error is a blocker for merging this PR, because I've used it a lot to diagnose download errors.

Did you also want advice about simplifying the code?

These changes are optional:

`Groth16Parameters::new` is getting really long. Can you split out the individual sapling and sprout downloads into separate methods?

If we do that change, we can simplify the loop to just handle the error case, because the success case doesn't need any extra code:
```rust
while let Err(error) = Groth16Parameters::download_sapling_parameters() {
    sapling_retries += 1;
    if sapling_retries >= PARAMETER_DOWNLOAD_MAX_RETRIES {
        panic!(
            "error downloading Sapling parameter files after {} retries. {:?} {}",
            PARAMETER_DOWNLOAD_MAX_RETRIES,
            error,
            Groth16Parameters::failure_hint(),
        );
    } else {
        tracing::info!(?error, "error downloading Zcash Sapling parameters, retrying");
    }
}
```

**Review by teor2345 — COMMENTED** [review]
Looks good, but we should check that the paths are the ones we expect, otherwise loading might fail. (This is a workaround for the `librustzcash` API, we should make it easier to use in future.)

See PR https://github.com/oxarbitrage/zebra/pull/162 for a commit that adds the asserts back in.

**Comment by teor2345** [comment]
The macOS failure is a peer set timeout, it should be fixed by PR #3292 or #3294.

**Review by teor2345 — APPROVED** [review]
Looks good, thanks!

**Pull request merged**